src/Http/Authentication.Core/src/AuthenticationHandlerProvider.cs

```diff
@@ -49,9 +49,22 @@ public AuthenticationHandlerProvider(IAuthenticationSchemeProvider schemes)
             {
                 return null;
             }
-            var handler = (context.RequestServices.GetService(scheme.HandlerType) ??
-                ActivatorUtilities.CreateInstance(context.RequestServices, scheme.HandlerType))
-                as IAuthenticationHandler;
+
+            var handler = (IAuthenticationHandler?) context.RequestServices.GetService(scheme.HandlerType);
+            if (handler == null)
+            {
+                if (scheme is AuthenticationSchemeWithFactory schemeWithFactory)
+                {
+                    handler = (IAuthenticationHandler) schemeWithFactory.ObjectFactory(context.RequestServices,
+                        Array.Empty<object>());
+                }
+                else
+                {
+                    handler = (IAuthenticationHandler) ActivatorUtilities.CreateInstance(context.RequestServices,
+                        scheme.HandlerType);
```

Comment:
Keeping the original soft casts to avoid invalid cast exceptions
```suggestion
                    handler = schemeWithFactory.ObjectFactory(context.RequestServices,
                        Array.Empty<object>()) as IAuthenticationHandler;
                }
                else
                {
                    handler = ActivatorUtilities.CreateInstance(context.RequestServices,
                        scheme.HandlerType) as IAuthenticationHandler;
```